proposals/0023-asynchronous-program-execution-and-broadcast.md

```diff
@@ -0,0 +1,371 @@
+as each validator's execution catches up to the slot and includes
+a BankHash confirming the state of the execution.
+
+It may be sufficient to wait for 1/3+ validators executing the state
+transition instead of the full 2/3+, because if 1/3+ are incorrect
+the cluster will halt anyways.
+
+### State synchronization
+
+Nodes must be able to compute a full snapshot at least once an
+epoch. So overall CU limits for the block must be set such that
+synchronous execution is possible. But because asynchronous execution
+is possible, replay can take advantage of much larger batches of
+execution.
+
+## Impact
+
+Multiple nodes can operate as Builder on the cluster concurrently.
+So clients can pick the nearest one, and the bandwidth to schedule
+and prioritize transactions is doubled.  There needs to be a design
+for BundleTransactions that allow the bundler to prioritize a batch
+of transactions for execution by paying a priority fee for all of
+them, and executing the whole batch together.
+
+Priority fees now also imply execution priority.
+
+Fork choice and voting is not blocked by user block propagation or
+execution of user transactions.
+
+Builders are going to capture all the MEV.
+
+Votes contain additional data, instead of BankHash, they contain
+(VoteHash, (BankHash, slot))
+
+## Security Considerations
+
+Validators may include incorrect BankHash values which may not
+be detected instantly.
+
+Validators may skip evaluating their own BankHash and copy results
+from other validators.
+
+Network halts if it cannot compute a epoch snapshot hash once an
+epoch because it is overloaded with user transactions.
+
+Leader could censor or delay the builder.
+
+Under heavy forking, validators will skip executing all the
+UserBlocks on minor forks.
+
+## Slashing UserBlock Builders
+
+There shouldn't be a reason to slash UserBlock builders.  Duplicate
+UserBlocks will be resolved by the leader, and it is up to the
+leader to pick one.  If turbine failed and the rest of the network
+has a different UserBlock, the leaders proposed block is dropped.
+
+If slashing is required, then stakeweighted or a bond based auction
+would be necessary.
+
+## Economic Considerations
+
+Builders should be shuffled and scheduled according to stake weight.
+
+TBD, deciding how should builders and leaders split the fees from
+user transactions.
+
+## Capacity Management
+
+Because capacity is split between UserBlocks, max UserBlock size
+as well as max account limit per UserBlock must be 1/N CUs, where
+N is the number of concurrent builders.
+
+To prevent under-utilization, unused resources from blocks that are
+partially filled can be used by future blocks up at most N * (block
+or account limit). This would allow the network to handle bursts
+of demand and on average maintain the desired load.
+
+## Multi leader spam
+
+Spammers would be motivated to send the same message to every
+leader.
+
+1. builders censor fee payers by most significant bit of the fee
```

Comment:
Am I missing something, shouldn't this be filtering by first `N` bits?

If we only look at first bit, then we'd get 2 sets of distinct fee-payers, rather than `N` (number of concurrent builders).

Reply:
if we use a power of 2 to limit the number of builders we can use ln(N) bits, as errror is always 0. Else using ln(N) + 4 is probably safer from a load shedding perspective error < 6.5%

Reply:
manipulating the fee payer is really annoying, for signature, I can just add a noop-instruction and rehash. why would we use fee payer and not signature?

Reply:
@apfitzge yea, I meant N bits

Reply:
@mschneider  for well behaving users, I think its more predictable to use a fee payer that is closest to the leader.  so latency sensitive users can rotate fee payers automatically